crates/masonry/src/text2/layout.rs

```diff
@@ -0,0 +1,463 @@
+// This software is licensed under Apache License 2.0 and distributed on an
+// "as-is" basis without warranties of any kind. See the LICENSE file for
+// details.
+
+//! A type for laying out, drawing, and interacting with text.
+
+use std::rc::Rc;
+
+use kurbo::{Affine, Line, Point, Rect, Size};
+use parley::fontique::{Style, Weight};
+use parley::layout::{Alignment, Cursor};
+use parley::style::{FontFamily, FontStack, GenericFamily, StyleProperty};
+use parley::{FontContext, Layout, LayoutContext};
+use vello::peniko::{Brush, Color};
+use vello::Scene;
+
+use super::{Link, TextStorage};
+
+/// A component for displaying text on screen.
+///
+/// This is a type intended to be used by other widgets that display text.
+/// It allows for the text itself as well as font and other styling information
+/// to be set and modified. It wraps an inner layout object, and handles
+/// invalidating and rebuilding it as required.
+///
+/// This object is not valid until the [`rebuild_if_needed`] method has been
+/// called. You should generally do this in your widget's [`layout`] method.
+/// Additionally, you should call [`needs_rebuild_after_update`]
+/// as part of your widget's [`update`] method; if this returns `true`, you will need
+/// to call [`rebuild_if_needed`] again, generally by scheduling another [`layout`]
+/// pass.
+///
+/// [`layout`]: trait.Widget.html#tymethod.layout
+/// [`update`]: trait.Widget.html#tymethod.update
+/// [`needs_rebuild_after_update`]: #method.needs_rebuild_after_update
+/// [`rebuild_if_needed`]: #method.rebuild_if_needed
+///
+/// TODO: Update docs to mentionParley
+#[derive(Clone)]
+pub struct TextLayout<T> {
+    text: T,
+    // TODO: Find a way to let this use borrowed data
+    scale: f32,
+
+    brush: Brush,
+    /// A brush used for handling disabled state without losing intentional state
+    override_brush: Option<Brush>,
+    font: FontStack<'static>,
+    text_size: f32,
+    weight: Weight,
+    style: Style,
+
+    alignment: Alignment,
+    max_advance: Option<f32>,
+
+    links: Rc<[(Rect, usize)]>,
+
+    needs_layout: bool,
+    needs_line_breaks: bool,
+    layout: Layout<Brush>,
+    layout_context: LayoutContext<Brush>,
+}
+
+/// Metrics describing the layout text.
+#[derive(Debug, Clone, Copy, Default)]
+pub struct LayoutMetrics {
+    /// The nominal size of the layout.
+    pub size: Size,
+    /// The distance from the nominal top of the layout to the first baseline.
+    pub first_baseline: f32,
+    /// The width of the layout, inclusive of trailing whitespace.
+    pub trailing_whitespace_width: f32,
+    //TODO: add inking_rect
+}
+
+impl<T> TextLayout<T> {
```

Comment:
Just a general comment that some things are prefixing with `set_` and others with `set_text_` and sometimes `set_color` is used in here and in other places `set_text_color` and it might be nice to make this more consistent. (Probably not immediately before RustNL...)

Reply:
Yeah, all of this API design could do with a tiny bit of cleanup; there's quite a lot of redudancy in these methods